peekingduck/viewer/viewer.py

```diff
@@ -0,0 +1,583 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implement PeekingDuck Viewer
+"""
+
+from typing import List
+from pathlib import Path
+import logging
+import platform
+import tkinter as tk
+from tkinter import filedialog
+from tkinter.messagebox import askyesno
+import threading
+import copy
+import cv2
+import numpy as np
+from PIL import Image, ImageTk
+from peekingduck.declarative_loader import DeclarativeLoader
+from peekingduck.pipeline.pipeline import Pipeline
+from peekingduck.viewer.playlist import PlayList
+from peekingduck.viewer.viewer_gui import create_window
+from peekingduck.viewer.viewer_utils import (
+    get_keyboard_char,
+    get_keyboard_modifier,
+)
+
+####################
+# Globals
+####################
+BUTTON_DELAY: int = 250  # milliseconds (0.25 of a second)
+BUTTON_REPEAT: int = int(1000 / 60)  # milliseconds (60 fps)
+CHANGE_PIPELINE_DELAY: float = 1.0  # seconds before switching pipelines
+FPS_60: int = int(1000 / 60)  # milliseconds per iteration
+ZOOM_TEXT: List[str] = ["50%", "75%", "100%", "125%", "150%", "200%", "250%", "300%"]
+ZOOM_DEFAULT_IDX: int = 2
+ZOOMS: List[float] = [0.5, 0.75, 1.0, 1.25, 1.50, 2.00, 2.50, 3.00]  # > 3x is slow!
+# PLAY_BUTTON_TEXT = "\u25B6"
+# STOP_BUTTON_TEXT = "\u23F9"
+# technotes: emoji-cons above don't render consistently on all OS'es, and/or play well
+# with some platforms (like Nvidia Jetsons)
+PLAY_BUTTON_TEXT = "Play"
+STOP_BUTTON_TEXT = "Stop"
+
+
+class Viewer:  # pylint: disable=too-many-instance-attributes, too-many-public-methods
+    """Implement PeekingDuck Viewer class"""
+
+    def __init__(
+        self,
+        pipeline_path: Path,
+        config_updates_cli: str,
+        custom_nodes_parent_subdir: str,
+        num_iter: int = 0,
+    ) -> None:
+        self.logger = logging.getLogger(__name__)
+        self.config_updates_cli = config_updates_cli
+        self.custom_nodes_parent_path = custom_nodes_parent_subdir
+        self.num_iter = num_iter
+        # init PlayList object
+        self.home_path = Path.home()
+        self.playlist = PlayList(self.home_path)
+        self.playlist.load_playlist_file()
+        self.playlist_show: bool = False
+        # init pipeline to run
+        self.init_pipeline(pipeline_path)
+        self.playlist.add_pipeline(self.pipeline_full_path)
+        # configure keyboard shortcuts map
+        self._keyboard_shortcuts = {
+            "z": self._zoom_reset,
+            "+": self._zoom_in,
+            "-": self._zoom_out,
+        }
```

Comment:
Not sure why `ctrl-z` is not working on Ubuntu: could it be trapped by the system?
Haven't addressed this, but one option is to replace it with `ctrl-=` instead.

Reply:
Could be some weird configuration on my system. If you have tried ctrl+z on your system and it works then it should be fine to just resolve this

Reply:
I have changed the zoom reset keyboard shortcut to `CTRL`-`=` instead. It should be better, since `CTRL-Z` is also the undo shortcut for Windows apps.